File: stash_preview.py
import json
import os
import re
import difflib
from typing import Tuple, Dict, List, Optional
from dataclasses import dataclass
from PIL import Image, ImageDraw
import logging

@dataclass
class ItemInfo:
    name: str
    slotId: int  # Changed from slot_id to match incoming JSON
    itemId: str  # Changed from item_id to match incoming JSON
    itemCount: int  # Changed from item_count to match incoming JSON

class ItemDataManager:
    def __init__(self):
        base_dir = os.path.dirname(os.path.abspath(__file__))
        assets_dir = os.path.join(base_dir, "assets")
        self.ITEM_DATA_FILE = os.path.join(assets_dir, "item-data.json")
        self.MATCHING_DB_FILE = os.path.join(assets_dir, "matchingdb.json")
        os.makedirs(assets_dir, exist_ok=True)
        
        self.item_data = self.load_json(self.ITEM_DATA_FILE)
        self.matching_db = self.load_matching_db()
        self.image_cache = {}

    @staticmethod
    def load_json(filename: str) -> dict:
        with open(filename, "r", encoding="utf-8") as f:
            return json.load(f)

    def load_matching_db(self) -> dict:
        try:
            return self.load_json(self.MATCHING_DB_FILE)
        except Exception:
            return {}

    def save_matching_db(self) -> None:
        with open(self.MATCHING_DB_FILE, "w", encoding="utf-8") as f:
            json.dump(self.matching_db, f, indent=2)

    @staticmethod
    def normalize_name(name: str) -> str:
        if not name:
            return ""
        name = name.lower()
        name = re.sub(r"designdataitem:|id_item_", "", name)
        name = re.sub(r"[^a-z0-9]", "", name)
        return name

    def get_item_image_path(self, item_name: str) -> Tuple[Optional[str], Optional[int], Optional[int], Optional[str]]:
        if item_name in self.matching_db:
            item_name = self.matching_db[item_name]

        norm_name = self.normalize_name(item_name)
        
        # Try exact match first
        for data in self.item_data.values():
            if self.normalize_name(data.get("name", "")) == norm_name:
                return (data["path"].replace("\\", os.sep),
                        data["inventory_width"],
                        data["inventory_height"],
                        data["name"])

        # Try fuzzy match as fallback
        name_map = {self.normalize_name(data.get("name", "")): data 
                   for data in self.item_data.values()}
        close = difflib.get_close_matches(norm_name, list(name_map.keys()), n=1, cutoff=0.7)
        
        if close:
            data = name_map[close[0]]
            logging.info(f"Fuzzy matched '{item_name}' → '{data['name']}'")
            return (data["path"].replace("\\", os.sep),
                    data["inventory_width"],
                    data["inventory_height"],
                    data["name"])
                    
        return None, None, None, None

class StashPreviewGenerator:
    def __init__(self, grid_width: int = 12, grid_height: int = 20, cell_size: int = 45):
        self.GRID_WIDTH = grid_width
        self.GRID_HEIGHT = grid_height
        self.CELL_SIZE = cell_size
        self.item_manager = ItemDataManager()
        logging.basicConfig(level=logging.INFO)

    def generate_preview(self, stash_id: str, items: List[ItemInfo]) -> Image.Image:
        preview = Image.new("RGBA", 
                          (self.GRID_WIDTH * self.CELL_SIZE, 
                           self.GRID_HEIGHT * self.CELL_SIZE))
        self._draw_grid(preview)
        
        for item in items:
            self._place_item(preview, item)
            
        return preview

    def _draw_grid(self, img: Image.Image) -> None:
        draw = ImageDraw.Draw(img)
        # Fill background
        draw.rectangle([0, 0, img.width, img.height], fill=(24, 20, 16, 255))
        # Draw border
        draw.rectangle([0, 0, img.width - 1, img.height - 1],
                      outline=(212, 175, 55, 255), width=4)
        # Draw grid lines
        grid_color = (60, 50, 30, 180)
        for x in range(1, self.GRID_WIDTH):
            x_pos = x * self.CELL_SIZE
            draw.line([(x_pos, 0), (x_pos, img.height)], fill=grid_color)
        for y in range(1, self.GRID_HEIGHT):
            y_pos = y * self.CELL_SIZE
            draw.line([(0, y_pos), (img.width, y_pos)], fill=grid_color)

    def _place_item(self, preview: Image.Image, item: ItemInfo) -> None:
        img_path, w, h, matched_name = self.item_manager.get_item_image_path(item.name)
        if not img_path or not os.path.exists(img_path):
            logging.warning(f"Item not found or missing image: {item.name}")
            return

        try:
            item_img = Image.open(img_path).convert("RGBA")
            expected_size = ((w or 1) * self.CELL_SIZE, (h or 1) * self.CELL_SIZE)
            if item_img.size != expected_size:
                item_img = item_img.resize(expected_size, Image.LANCZOS)
            
            x, y = item.slotId % self.GRID_WIDTH, item.slotId // self.GRID_WIDTH  # Updated to use slotId
            preview.paste(item_img, (x * self.CELL_SIZE, y * self.CELL_SIZE), item_img)
            logging.info(f"Placed '{matched_name}' at ({x},{y})")
            
            # Record successful match
            self.item_manager.matching_db[item.name] = matched_name
            
        except Exception as e:
            logging.error(f"Failed to process image {img_path}: {e}")

def get_item_name_from_id(item_id, item_data):
    # Extract base name from item_id
    if not item_id.startswith("DesignDataItem:Id_Item_"):
        return None
    base = item_id[len("DesignDataItem:Id_Item_"):]
    # Remove trailing _xxxx numbers
    base = re.sub(r'_\d+$', '', base)
    # Replace underscores with spaces
    base = base.replace("_", " ")
    return base

def parse_stashes(packet_data, item_data):
    stashes = {}
    storage_infos = packet_data.get("characterDataBase", {}).get("CharacterStorageInfos", [])
    for storage in storage_infos:
        inventory_id = storage.get("inventoryId")
        items = storage.get("CharacterStorageItemList", [])
        stash_items = []
        for item in items:
            item_id = item.get("itemId", "")
            slot_id = item.get("slotId")
            if slot_id is None:
                continue
            name = get_item_name_from_id(item_id, item_data)
            stash_items.append({
                "name": name,
                "slotId": slot_id,  # Already matches ItemInfo field name
                "itemId": item_id,   # Already matches ItemInfo field name
                "itemCount": item.get("itemCount", 1)  # Already matches ItemInfo field name
            })
        if stash_items:
            stashes[inventory_id] = stash_items
    return stashes

def slotid_to_xy(slot_id):
    return slot_id % 12, slot_id // 12

def main():
    base_dir = os.path.dirname(os.path.abspath(__file__))
    packet_data_path = os.path.join(base_dir, "packet_data.json")
    output_dir = os.path.join(base_dir, "output")
    os.makedirs(output_dir, exist_ok=True)

    try:
        item_data = ItemDataManager().item_data
        if not os.path.exists(packet_data_path):
            print(f"Error: {packet_data_path} not found. Please run packet capture first.")
            return
            
        packet_data = ItemDataManager.load_json(packet_data_path)
        matching_db = {}  # collect original → matched names

        stashes = parse_stashes(packet_data, item_data)
        if not stashes:
            print("No stashes found in packet data.")
            return

        generator = StashPreviewGenerator()

        for stash_id, items in stashes.items():
            print(f"\nProcessing stash inventoryId={stash_id} with {len(items)} items...")
            preview = generator.generate_preview(stash_id, [ItemInfo(**item) for item in items])
            outname = os.path.join(output_dir, f"stash_preview_{stash_id}.png")
            preview.save(outname)
            print(f"Preview saved as {outname}")

        # save the matching database
        generator.item_manager.save_matching_db()
        print("Matching DB saved as matchingdb.json")
        
    except Exception as e:
        print(f"Error generating previews: {e}")
        logging.error(f"Failed to generate previews: {e}", exc_info=True)

if __name__ == "__main__":
    main()

File: nodejs/stashPreview.js
const sharp = require('sharp');
const fs = require('fs-extra');
const path = require('path');

class ItemInfo {
    constructor({ name, slotId, itemId, itemCount }) {
        this.name = name;
        this.slotId = slotId;
        this.itemId = itemId;
        this.itemCount = itemCount || 1;
    }
}

class ItemDataManager {
    constructor() {
        // Resolve project root one level above this script
        const scriptDir = path.dirname(__filename);
        this.baseDir = path.resolve(scriptDir, "..");
        this.assetsDir = path.join(this.baseDir, "assets");
        this.ITEM_DATA_FILE = path.join(this.assetsDir, "item-data.json");
        this.MATCHING_DB_FILE = path.join(this.assetsDir, "matchingdb.json");
        fs.ensureDirSync(this.assetsDir);

        this.itemData = this.loadJson(this.ITEM_DATA_FILE);
        this.matchingDb = this.loadMatchingDb();
        this.imageCache = new Map();
        this.unmatchedItems = new Set();
        this.itemStats = {
            matched: new Set(),
            unmatched: new Set(),
            modified: new Set()
        };
    }

    loadJson(filename) {
        return fs.readJsonSync(filename, { encoding: 'utf-8' });
    }

    loadMatchingDb() {
        try {
            return this.loadJson(this.MATCHING_DB_FILE);
        } catch {
            return {};
        }
    }

    saveMatchingDb() {
        fs.writeJsonSync(this.MATCHING_DB_FILE, this.matchingDb, { spaces: 2 });
    }

    normalizeName(name) {
        if (!name) return "";
        name = name.replace("DesignDataItem:Id_Item_", "");
        name = name.replace(/[\s'-]/g, "");
        return name.toLowerCase();
    }

    async getItemImagePath(itemName) {
        // Check matching DB first
        if (itemName in this.matchingDb) {
            const matchedName = this.matchingDb[itemName];
            const data = Object.values(this.itemData).find(d => d.name === matchedName);
            if (data) {
                this.itemStats.matched.add(itemName);
                const rel = data.path.replace(/\\/g, path.sep);
                const full = path.join(this.baseDir, rel);
                return { path: full, width: data.inventory_width, height: data.inventory_height, name: data.name };
            }
        }

        // Try direct match
        const normName = this.normalizeName(itemName);
        const data = Object.values(this.itemData).find(d => this.normalizeName(d.name) === normName);

        if (data) {
            if (!(itemName in this.matchingDb)) {
                this.itemStats.modified.add(itemName);
            }
            const rel = data.path.replace(/\\/g, path.sep);
            const full = path.join(this.baseDir, rel);
            return { path: full, width: data.inventory_width, height: data.inventory_height, name: data.name };
        }

        // Track unmatched items
        this.itemStats.unmatched.add(itemName);
        if (!(itemName in this.matchingDb)) {
            this.unmatchedItems.add(itemName);
        }
        return null;
    }

    saveUnmatchedItems() {
        if (this.unmatchedItems.size > 0) {
            const timestamp = new Date().toISOString().replace(/[:.]/g, '-');
            const filename = path.join(path.dirname(this.MATCHING_DB_FILE), 
                `unmatched_items_${timestamp}.txt`);
            
            const content = Array.from(this.unmatchedItems).sort().join('\n');
            fs.writeFileSync(filename, content, 'utf-8');
            
            console.log(`\nItem Processing Statistics:`);
            console.log(`- Matched Items: ${this.itemStats.matched.size}`);
            console.log(`- Modified Items: ${this.itemStats.modified.size}`);
            console.log(`- Unmatched Items: ${this.itemStats.unmatched.size}`);
            console.log(`\nSaved ${this.unmatchedItems.size} unmatched items to ${filename}`);
        }
    }
}

class StashPreviewGenerator {
    constructor(gridWidth = 12, gridHeight = 20, cellSize = 45) {
        this.GRID_WIDTH = gridWidth;
        this.GRID_HEIGHT = gridHeight;
        this.CELL_SIZE = cellSize;
        this.itemManager = new ItemDataManager();
    }

    async generatePreview(stashId, items) {
        const width = this.GRID_WIDTH * this.CELL_SIZE;
        const height = this.GRID_HEIGHT * this.CELL_SIZE;

        console.log(`\nProcessing ${items.length} items in stash ${stashId}...`);

        // Initialize Sharp image with background
        const image = sharp({ create: { width, height, channels: 4, background: { r:24, g:20, b:16, alpha:255 } } });

        // Collect overlays: grid and items
        const composites = [];
        // Grid overlay
        const gridSVG = this._createGridSVG(width, height, { r:60, g:50, b:30, alpha:180 }, { r:212, g:175, b:55, alpha:255 });
        composites.push({ input: Buffer.from(gridSVG), top: 0, left: 0 });

        for (const item of items) {
            const itemData = await this.itemManager.getItemImagePath(item.name);
            if (!itemData) {
                console.warn(`  ❌ No data entry for item: ${item.name}`);
                continue;
            }
            const imgPath = itemData.path;
            if (!fs.existsSync(imgPath)) {
                console.warn(`  ❌ File not found for item '${item.name}': ${imgPath}`);
                continue;
            }
            try {
                const w = (itemData.width || 1) * this.CELL_SIZE;
                const h = (itemData.height || 1) * this.CELL_SIZE;
                const buf = await sharp(imgPath).resize(w, h).toBuffer();
                const x = item.slotId % this.GRID_WIDTH;
                const y = Math.floor(item.slotId / this.GRID_WIDTH);
                console.log(`  ✓ Queue '${itemData.name}' at (${x},${y})`);
                this.itemManager.matchingDb[item.name] = itemData.name;
                composites.push({ input: buf, top: y * this.CELL_SIZE, left: x * this.CELL_SIZE });
            } catch (e) {
                console.error(`  ❌ Failed to process ${imgPath}:`, e);
            }
        }

        // Composite grid and items on background
        return image.composite(composites);
    }

    _createGridSVG(width, height, gridColor, borderColor) {
        const lines = [];
        
        // Vertical lines
        for (let x = 1; x < this.GRID_WIDTH; x++) {
            const xPos = x * this.CELL_SIZE;
            lines.push(`<line x1="${xPos}" y1="0" x2="${xPos}" y2="${height}" 
                stroke="rgba(${gridColor.r},${gridColor.g},${gridColor.b},${gridColor.alpha/255})" 
                stroke-width="1" shape-rendering="crispEdges" />`);
        }
        
        // Horizontal lines
        for (let y = 1; y < this.GRID_HEIGHT; y++) {
            const yPos = y * this.CELL_SIZE;
            lines.push(`<line x1="0" y1="${yPos}" x2="${width}" y2="${yPos}" 
                stroke="rgba(${gridColor.r},${gridColor.g},${gridColor.b},${gridColor.alpha/255})" 
                stroke-width="1" shape-rendering="crispEdges" />`);
        }

        // Border
        const border = `<rect x="0" y="0" width="${width}" height="${height}" fill="none" 
            stroke="rgba(${borderColor.r},${borderColor.g},${borderColor.b},${borderColor.alpha/255})" 
            stroke-width="4" shape-rendering="crispEdges" />`;

        return `<svg width="${width}" height="${height}" xmlns="http://www.w3.org/2000/svg" 
                shape-rendering="crispEdges">
            ${lines.join('\n')}
            ${border}
        </svg>`;
    }
}

function getItemNameFromId(itemId) {
    if (!itemId.startsWith("DesignDataItem:Id_Item_")) {
        return null;
    }
    let base = itemId.substring("DesignDataItem:Id_Item_".length);
    base = base.replace(/_\d+$/, '');
    base = base.replace(/_/g, ' ');
    return base;
}

function parseStashes(packetData, itemData) {
    const stashes = {};
    const storageInfos = packetData?.characterDataBase?.CharacterStorageInfos || [];

    for (const storage of storageInfos) {
        const inventoryId = storage.inventoryId;
        const items = storage.CharacterStorageItemList || [];
        const stashItems = [];
        const usedSlots = new Set();

        // Process items with defined slots
        for (const item of items) {
            if ('slotId' in item) {
                const itemId = item.itemId || '';
                const slotId = item.slotId;
                const name = getItemNameFromId(itemId);
                stashItems.push({
                    name,
                    slotId,
                    itemId,
                    itemCount: item.itemCount || 1
                });
                usedSlots.add(slotId);
            }
        }

        // Process items without slots
        for (const item of items) {
            if (!('slotId' in item)) {
                const itemId = item.itemId || '';
                let slotId = 0;
                while (usedSlots.has(slotId)) {
                    slotId++;
                }
                const name = getItemNameFromId(itemId);
                stashItems.push({
                    name,
                    slotId,
                    itemId,
                    itemCount: item.itemCount || 1
                });
                usedSlots.add(slotId);
            }
        }

        if (stashItems.length > 0) {
            stashes[inventoryId] = stashItems;
        }
    }
    return stashes;
}

async function main() {
    // Use project root as base directory
    const scriptDir = path.dirname(__filename);
    const baseDir = path.resolve(scriptDir, "..");
    const packetDataPath = path.join(baseDir, "packet_data.json");
    const outputDir = path.normalize(path.join(baseDir, "output"));

    try {
        // Ensure output directory exists
        await fs.ensureDir(outputDir);
        
        // Test write permissions by creating a test file
        const testFile = path.join(outputDir, '.test');
        await fs.writeFile(testFile, '');
        await fs.remove(testFile);

        const itemDataManager = new ItemDataManager();
        
        if (!fs.existsSync(packetDataPath)) {
            console.error(`Error: ${packetDataPath} not found. Please run packet capture first.`);
            return;
        }

        const packetData = await fs.readJson(packetDataPath);
        const stashes = parseStashes(packetData, itemDataManager.itemData);

        if (Object.keys(stashes).length === 0) {
            console.log("No stashes found in packet data.");
            return;
        }

        console.log("\nStarting stash preview generation...");
        console.log(`Output directory: ${outputDir}`);
        
        const generator = new StashPreviewGenerator();

        for (const [stashId, items] of Object.entries(stashes)) {
            const preview = await generator.generatePreview(
                stashId, 
                items.map(item => new ItemInfo(item))
            );
            
            const outname = path.join(outputDir, `stash_preview_${stashId}.png`);
            await preview.toFile(outname);
            console.log(`✓ Preview saved as ${outname}`);
        }

        generator.itemManager.saveUnmatchedItems();
        generator.itemManager.saveMatchingDb();
        console.log("\n✓ Matching DB saved as matchingdb.json");

    } catch (e) {
        console.error("\n❌ Error generating previews:", e);
        if (e.code === 'EACCES') {
            console.error(`Permission denied accessing output directory: ${outputDir}`);
        }
    }
}

if (require.main === module) {
    main().catch(console.error);
}

module.exports = {
    StashPreviewGenerator,
    ItemInfo,
    ItemDataManager,
    parseStashes,
    getItemNameFromId
};